homeassistant/components/wittiot/sensor.py

```diff
@@ -0,0 +1,558 @@
+            sensors.append(
+                MainDevWittiotSensor(
+                    coordinator,
+                    entry.data[CONF_HOST],
+                    entry.data[DEVICE_NAME],
+                    desc,
+                )
+            )
+
+    # Subdevice Data
+    for key in coordinator.data:
+        if coordinator.data.get(key) not in ("", "--", "--.-", "None"):
+            if key in MultiSensorInfo.SENSOR_INFO:
+                mapping = WITTIOT_SENSORS_MAPPING[
+                    MultiSensorInfo.SENSOR_INFO[key]["data_type"]
+                ]
+                description = dataclasses.replace(
+                    mapping,
+                    key=key,
+                    sensor_type=MultiSensorInfo.SENSOR_INFO[key]["dev_type"],
+                    name=MultiSensorInfo.SENSOR_INFO[key]["name"],
+                )
+                sensors.append(
+                    SubDevWittiotSensor(
+                        coordinator,
+                        entry.data[CONF_HOST],
+                        entry.data[DEVICE_NAME],
+                        description,
+                    )
+                )
+
+    async_add_entities(sensors, True)
+
+
+class MainDevWittiotSensor(
+    CoordinatorEntity[WittiotDataUpdateCoordinator], SensorEntity
+):
+    """Define an Local sensor."""
+
+    _attr_has_entity_name = True
+    entity_description: WittiotSensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: WittiotDataUpdateCoordinator,
+        host: str,
+        devname: str,
+        description: WittiotSensorEntityDescription,
+    ) -> None:
+        """Initialize."""
+        super().__init__(coordinator)
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{devname}_{MAIN_DATA}")},
+            manufacturer="WittIOT",
+            name=f"{devname}_{MAIN_DATA}",
+            model=self.coordinator.data["ver"],
+            configuration_url=f"http://{host}",
+        )
+        self._attr_unique_id = f"{devname}_{description.key}"
+        self.entity_description = description
+
+    @property
+    def native_value(self) -> str | int | float | None:
+        """Return the state."""
+        state = self.coordinator.data[self.entity_description.key]
+        if state not in ("", "--", "--.-", "None"):
+            return state
+        return None
+
+
+class SubDevWittiotSensor(
+    CoordinatorEntity[WittiotDataUpdateCoordinator], SensorEntity
+):
+    """Define an Local sensor."""
+
+    _attr_has_entity_name = True
+    entity_description: WittiotSensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: WittiotDataUpdateCoordinator,
+        host: str,
+        devname: str,
+        description: WittiotSensorEntityDescription,
+    ) -> None:
+        """Initialize."""
+        super().__init__(coordinator)
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{devname}_{description.sensor_type}")},
```

Comment:
where have the channels gone?

Reply:
Moved to the library to determine.

Reply:
How do we handle different channels?

Reply:
In the library, I store the channel information in an object, and after receiving the returned data, I use this object to distinguish the data contained in different channels. For example, soil moisture CH1 contains the moisture and electricity of CH1, which are displayed as a sub-device.